### PR TITLE
Freeze initial reserve gold

### LINKED ISSUE #2390: Initial reserve cGLD transferability
### Expected Behavior

With X cGLD allocated to the reserve at launch, only Y cGLD should be transferable by reserve spenders  immediately after launch (for rebalancing purposes) whereas the rest, X-Y cGLD, should be frozen and become unfrozen linearly over the course of Z days. The parameters X, Y and Z are tbd. 

Additionally, similar to `getReserveGoldBalance()`, there should be a function `getUnfrozenReserveGoldBalance()` that returns the sum of the currently unfrozen reserve balance and the balances of the `otherReserveAddresses`.

### Current Behavior

All X cGLD allocated to the reserve are, ignoring spending limits, immediately transferable after launch.

### PR DESCRIPTION
**Pull request by jfoutts-celo**
### Description

Freezes the initial reserve gold according to a linear parameterization.

### Tested

Added unit tests.

### Related issues

- Fixes #2390

### CODE REVIEW

**Comment by jfoutts-celo** [comment]
Will the reserve already have it's initial balance at the time of `initialize()`?

This relies on block.timestamp (as does the spending limits which I copied).  Would it be better to use block.number?

File: packages/protocol/contracts/stability/Reserve.sol
Comment (rcroessmann) [inline packages/protocol/contracts/stability/Reserve.sol]: Nitty: This is also found in getReserveGoldBalance() and for readability it might be worth to create a separate function getOtherReserveAddressesGoldBalance() and to then use this function in getUnfrozenReserveGoldBalance() and getReserveGoldBalance().

**Comment by rcroessmann** [comment]
This looks good to me but would love to have a review from you @asaj as well as I am still pretty new to this stuff. 

Also, I left a comment in https://github.com/celo-org/celo-monorepo/pull/2395 a while back that I think applies now:  "If we decide to do the above, we should make sure that the unfrozenReserveGoldBalance is used instead of reserveGoldBalance in the reserve ratio calculation as soon as #2390 is tackled." Should we create an issues for that?

File: packages/protocol/contracts/stability/Reserve.sol
Comment (jfoutts-celo) [reply packages/protocol/contracts/stability/Reserve.sol]: Good call. Updated.

**Comment by jfoutts-celo** [comment]
Should create an issue for more discussion?

If we're sure that `computeTobinTax()` should use `getUnfrozenReserveGoldBalance()` instead of `getReserveGoldBalance()` then we can put into this PR.

But we'll need to be careful with the parameters, as they could trigger Tobin tax right after launch.

**Comment by asaj** [comment]
> If we're sure that computeTobinTax() should use getUnfrozenReserveGoldBalance() instead of getReserveGoldBalance() then we can put into this PR.

This seems like the right approach to me

**Comment by rcroessmann** [comment]
> > If we're sure that computeTobinTax() should use getUnfrozenReserveGoldBalance() instead of getReserveGoldBalance() then we can put into this PR.
> 
> This seems like the right approach to me

+1 on adding it to this PR

**Comment by rcroessmann** [comment]
Maybe also worth putting into this PR are two other things that should be updated to use unfrozenReserveGoldBalance. 
1. The exchange cGLD tank size: Currently, tanks are set as a fraction of the reserveGoldBalance. This could lead to a situation where the exchange contract tries to transfer more cGLD to a user as is currently unfrozen. So it seems like we would need to use the unfrozenReserveGoldBalance in this line https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/Exchange.sol#L303
2. The spending limit: Reserve spenders can currently still (try to) spend x% of the reserveGoldBalance each day but the pct. should be applied to the unfrozenReserveGoldBalance. https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/Reserve.sol#L241

File: packages/protocol/contracts/stability/Reserve.sol
Comment (asaj) [inline packages/protocol/contracts/stability/Reserve.sol]: nit: use safemath

File: packages/protocol/contracts/stability/Reserve.sol
Comment (asaj) [inline packages/protocol/contracts/stability/Reserve.sol]: ditto

File: packages/protocol/contracts/stability/Reserve.sol
Comment (asaj) [inline packages/protocol/contracts/stability/Reserve.sol]: ditto

File: packages/protocol/contracts/stability/Reserve.sol
Comment (asaj) [inline packages/protocol/contracts/stability/Reserve.sol]: ditto

File: packages/protocol/contracts/stability/Reserve.sol
Comment (asaj) [inline packages/protocol/contracts/stability/Reserve.sol]: ditto

File: packages/protocol/migrationsConfig.js
Comment (asaj) [inline packages/protocol/migrationsConfig.js:99]: @rcroessmann what do you suggest for these values?

**Review by asaj — APPROVED** [review]
LGTM pending changes

File: packages/protocol/contracts/stability/Reserve.sol
Comment (jfoutts-celo) [reply packages/protocol/contracts/stability/Reserve.sol]: Thanks. I forgot there's no operator overloads in Solidity.  Created an issue to add a SafeMath always lint: https://github.com/celo-org/celo-monorepo/issues/2810

**Review by rcroessmann — APPROVED** [review]
LGTM, pending failed tests.

**Pull request merged**
